### ISSUE TITLE
Memory copy loop optimization

### ISSUE THREAD

**Issue by k06a** [wighawag/clones-with-immutable-args#2]
This loop could be replaced with CALLDATACOPY():

https://github.com/wighawag/clones-with-immutable-args/blob/5950723ffcfa047f13262e5dbd7218b54360c42e/src/ClonesWithImmutableArgs.sol#L126-L134

**Comment by hrkrshnn** [comment]
Alternatively, if one really wants to keep the data in memory, one can use the identity precompile to copy from memory to memory. It should be cheaper if that loop runs more than 2 times (please check).

**Comment by k06a** [comment]
Yep, I see method is internal and argument is not calldata, then really consider using “identity” precompile for copying

**Comment by WyseNynja** [comment]
If that loop is replaced, copyPtr and dataPtr both need to be increased still, right?

**Comment by z0r0z** [comment]
this looks great. any thoughts on integrating @wighawag ?

**Comment by 0xca11** [comment]
Implemented this optimization using the identity precompile on this fork in case anyone's interested: https://github.com/RollaProject/clones-with-immutable-args/commit/3cef7aeca219cdfb3370c3e7a59f40c022fc89ae

@z0r0z @wighawag Please let me know if a PR with this change would be welcome :)

**Comment by WyseNynja** [comment]
How much did deploy and call costs change with the identity precompile?

**Comment by wighawag** [comment]
Took the safe part of this in <commit>: the copy is now a single Yul loop, and the mask on the trailing partial word turned out to be redundant, since those bytes are always overwritten by the length suffix that follows. That also removed an out of bounds read when data.length was a multiple of 32.

Measured, that is 25 gas at 64 byte args and 115 at 1024.

Not doing the identity precompile, and not doing MCOPY either. At 1024 bytes of args the whole copy is around 0.3% of the cost of the clone, because code deposit at 200 gas per byte dominates. The precompile also carries a ~100 gas call overhead, so its break even is much further out than 2 iterations. MCOPY would mean requiring Cancun and Solidity 0.8.25 from every consumer of a ^0.8.4 library for a fraction of a percent.

Thanks for the original suggestion, it pointed at the right code even if the answer ended up being a different one.
